crates/bevy_math/src/bounding/raytest3d.rs

```diff
@@ -0,0 +1,85 @@
+use super::{Aabb3d, BoundingSphere, IntersectsVolume};
+use crate::{primitives::Direction3d, Vec3};
+
+/// A raycast intersection test for 3D bounding volumes
+pub struct RayTest3d {
+    /// The origin of the ray
+    pub origin: Vec3,
+    /// The direction of the ray
+    pub dir: Direction3d,
+    /// The inverse direction of the ray
+    pub inv_dir: Vec3,
+    /// The maximum time of impact of the ray
+    pub max: f32,
+}
+
+impl RayTest3d {
+    /// Construct a [`RayTest3d`] from an origin, [`Direction3d`] and max time of impact.
+    pub fn new(origin: Vec3, dir: Direction3d, max: f32) -> Self {
+        Self {
+            origin,
+            inv_dir: Vec3::ONE / *dir,
+            dir,
+            max,
+        }
+    }
+
+    /// Get the time of impact for an intersection with an [`Aabb3d`], if any.
+    pub fn aabb_intersection_at(&self, aabb: &Aabb3d) -> Option<f32> {
+        let (min_x, max_x) = if self.inv_dir.x.is_sign_positive() {
+            (aabb.min.x, aabb.max.x)
+        } else {
+            (aabb.max.x, aabb.min.x)
+        };
+        let (min_y, max_y) = if self.inv_dir.y.is_sign_positive() {
+            (aabb.min.y, aabb.max.y)
+        } else {
+            (aabb.max.y, aabb.min.y)
+        };
+        let (min_z, max_z) = if self.inv_dir.z.is_sign_positive() {
+            (aabb.min.z, aabb.max.z)
+        } else {
+            (aabb.max.z, aabb.min.z)
+        };
+        let tmin_x = (min_x - self.origin.x) * self.inv_dir.x;
+        let tmin_y = (min_y - self.origin.y) * self.inv_dir.y;
+        let tmin_z = (min_z - self.origin.z) * self.inv_dir.z;
+        let tmax_x = (max_x - self.origin.x) * self.inv_dir.x;
+        let tmax_y = (max_y - self.origin.y) * self.inv_dir.y;
+        let tmax_z = (max_z - self.origin.z) * self.inv_dir.z;
+
+        let tmin = tmin_x.max(tmin_y).max(tmin_z).max(0.);
+        let tmax = tmax_z.min(tmax_y).min(tmax_x).min(self.max);
+
+        if tmin <= tmax {
+            Some(tmin)
+        } else {
+            None
+        }
+    }
+
+    /// Get the time of impact for an intersection with a [`BoundingSphere`], if any.
+    pub fn sphere_intersection_at(&self, sphere: &BoundingSphere) -> Option<f32> {
```

Comment:
Working through the math and guessing the variable names:
(using <|> notation to denote scalar products)
```
offset = origin - center
offset_projected_on_ray = <offset|dir>

closest_point = origin - offset_projected_on_ray * dir
closest_point_offset = closest_point - center = offset - offset_projected_on_ray * dir

closest_point_distance_squared = <closest_point_offset|closest_point_offset>

h = radius * radius - closest_point_distance_squared
```

if `h < 0`, then the closest point is outside the sphere, therefore we missed => `return None`
if `h > 0`, then the closest point is inside the sphere, therefore we have a line hit
no idea if comparing `h > -offset_projected_on_ray` is the right thing?
assuming `origin` is outside the sphere, then it should be fine to compare `offset_projected_on_ray < 0`, meaning the linehit is in the positive part of the line, not behind where we shoot the ray from.
if `origin` is inside the sphere... well that's a ray hit I believe?

Let's assume we hit the sphere:
```
col_time = -offset_projected_on_ray - h
hit_point = origin + col_time * dir
```
We expect the distance `|hit_point - center|` to be less than (ideally equal to) the radius, therefore:
```
|hit_point - center| = |origin + col_time * dir - center| = |offset + col_time * dir| = 
|offset - (offset_projected_on_ray + h) * dir| = |closest_point_offset - h * dir| = 
|closest_point_offset - (radius * radius - closest_point_distance_squared) * dir|
```
Next we use that `|a|^2 = <a|a>` and that `closest_point_offset` and `dir` are perpendicular so their product is 0.
```
|closest_point_offset - (radius * radius - closest_point_distance_squared) * dir|^2 = 
|closest_point_offset|^2 - |(radius * radius - closest_point_distance_squared) * dir|^2 =
closest_point_distance_squared - (radius^2 - closest_point_distance_squared)^2 =... doesn't equal radius^2
```
I believe there is a missing `sqrt` in the calculation of the hit time. (One can also guess this from unit analysis, `h` is distance^2 cause it involves `radius`^2 and `qc`^2 while `b` is just proportional to distance, because `dir` is normalised).

Reply:
Yea I think you're right about the missing `sqrt`. I based the code on this: https://iquilezles.org/articles/intersectors/ (the second example at sphere iirc), and there's clearly a `h = sqrt(h)` I forgot to port. I'm not sure if your guesses on the names are right, I'll see if I can rename them after I've tested these sphere intersections.